src/main/java/uk/gov/ida/verifyserviceprovider/services/UnsignedAssertionsResponseHandler.java

```diff
@@ -0,0 +1,129 @@
+import uk.gov.ida.saml.security.EidasValidatorFactory;
+import uk.gov.ida.saml.security.SecretKeyDecryptorFactory;
+import uk.gov.ida.saml.security.exception.SamlFailedToDecryptException;
+import uk.gov.ida.saml.security.validators.ValidatedResponse;
+import uk.gov.ida.saml.security.validators.encryptedelementtype.EncryptionAlgorithmValidator;
+import uk.gov.ida.verifyserviceprovider.validators.InstantValidator;
+
+import javax.crypto.NoSuchPaddingException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import static uk.gov.ida.saml.security.errors.SamlTransformationErrorFactory.unableToDecrypt;
+
+public class UnsignedAssertionsResponseHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(UnsignedAssertionsResponseHandler.class);
+
+    private final EidasValidatorFactory eidasValidatorFactory;
+    private final StringToOpenSamlObjectTransformer<Response> stringToResponseTransformer;
+    private final InstantValidator instantValidator;
+    private final SecretKeyDecryptorFactory secretKeyDecryptorFactory;
+    private final EncryptionAlgorithmValidator encryptionAlgorithmValidator;
+
+    public UnsignedAssertionsResponseHandler (
+            EidasValidatorFactory eidasValidatorFactory,
+            StringToOpenSamlObjectTransformer<Response> stringToResponseTransformer,
+            InstantValidator instantValidator,
+            SecretKeyDecryptorFactory secretKeyDecryptorFactory,
+            EncryptionAlgorithmValidator encryptionAlgorithmValidator
+    ) {
+        this.eidasValidatorFactory = eidasValidatorFactory;
+        this.stringToResponseTransformer = stringToResponseTransformer;
+        this.instantValidator = instantValidator;
+        this.secretKeyDecryptorFactory = secretKeyDecryptorFactory;
+        this.encryptionAlgorithmValidator = encryptionAlgorithmValidator;
+    }
+
+    public ValidatedResponse getValidatedResponse(
+            Assertion hubResponseAssertion,
+            String expectedInResponseTo
+    ) {
+        ValidatedResponse validatedResponse = eidasValidatorFactory.getValidatedResponse(
+                stringToResponseTransformer.apply(
+                        getCountryResponseStringFromAssertion(hubResponseAssertion)
+                )
+        );
+
+        if (!expectedInResponseTo.equals(validatedResponse.getInResponseTo())) {
+            throw new SamlResponseValidationException(
+                    String.format("Expected InResponseTo to be %s, but was %s", expectedInResponseTo, validatedResponse.getInResponseTo())
+            );
+        }
+
+        instantValidator.validate(validatedResponse.getIssueInstant(), "Response IssueInstant");
+
+        return validatedResponse;
+    }
+
+    public List<Assertion> decryptAssertion(
+            ValidatedResponse validatedResponse,
+            Assertion hubResponseAssertion
```

Comment:
I might have called this parameter `hubResponseAssertionWithEncryptedDecryptionKeys` or something similar. If I understand correctly, it contains a number of keys that can be used to decrypt the assertions.

Reply:
So this is the assertion from the hub response, which contains the original country saml response, as well as the encrypted decryption keys. Could be a bit of a mouthful to get it all in.